README.md
# Spawn
> Swate - something or someone that gets you absolutely joyed (Urban dictionary)

__Spawn__ is a **S**wate **P**rotocol **A**rchitect for **W**ord **N**otes.

Spawn tries to make the transfer of old Word protocols into modern workflow annotations as easy and fast as possible. We will design this process to be perfectly integrable with [Swate](https://github.com/nfdi4plants/Swate) and [Arc](https://github.com/nfdi4plants/ARC).

## Install/Use

You can test Spawn by following the steps below. If you decide to do so, please take the time to set up a Github account to report your issues and suggestions [here](https://github.com/nfdi4plants/Spawn/issues).
You can also search existing issues for solutions for your questions and/or discussions about your suggestions.

### Word desktop (WIP)

#### Using the release archive

- Install [node.js LTS](https://nodejs.org/en/) (needed for office addin related tooling)
- Download the [latest test release archive]() and extract it
- Execute the test.cmd (windows, as administrator) or test.sh (macOS, you will need to make it executable via chmod a+x) script.

#### Using a shared folder 

If you have administrative access in your organization, you can create a network share folder and follow [this guide](https://github.com/OfficeDev/office-js-docs-pr/blob/master/docs/testing/create-a-network-shared-folder-catalog-for-task-pane-and-content-add-ins.md#:~:text=Sideload%20your%20add%2Din,-Put%20the%20manifest&text=Be%20sure%20to%20specify%20the,element%20of%20the%20manifest%20file.&text=In%20Excel%2C%20Word%2C%20or%20PowerPoint,Office%20Add%2Dins%20dialog%20box.) to make the addin available without any further downloads from your users

### Excel online (WIP)

- Download the [latest test release archive]() and extract it
- Launch Excel online, open a (blank) workbook 
- Under the `Insert` tab, select `Add-Ins`
- Go to `Manage my Add-Ins` and select `Upload my Add-In`
- select and upload the `manifest.xml` file contained in the test release archive.

## Discuss/Get help

### Documentation (WIP)

#### API docs

The RPC API docs of Spawn are available [here]().

#### Feature documentation

Documentation of the features is now a top priority as we completed the POC stage :rocket:. You can discuss needed docs and the available ones [here]()

#### Guides

Guides will be coming in the near future.

### Report problems and request features

Please use [Github issues](https://github.com/nfdi4plants/Spawn/issues/new) to track problems, feature requests, questions, and discussions. Additionally, you can get in touch with us on [Twitter](https://twitter.com/nfdi4plants)

### (Beta) Discuss in the GitHub Discussion forum (WIP)

GitHub recently added forum-like features to repositories, which makes it possible to discuss and share things you do with swate without creating issues. You are very welcome to post content [in the `Discussions` tab](), but please use issues for bug reports/feature requests (or, if some of those result from a discussion, create a new issue referencing the discussion)

## Develop

### Contribute

Before you contribute to the project remember to return all placeholders to your project:

-   webpack.config.js    
    ```
    https: {
        key: "{USERFOLDER}/.office-addin-dev-certs/localhost.key",
        cert: "{USERFOLDER}/.office-addin-dev-certs/localhost.crt",
        ca: "{USERFOLDER}/.office-addin-dev-certs/ca.crt"
    },
    ```

### Prerequisites

 - .NET Core SDK at of at least the version in [the global.json file](global.json)
 - Node.js with npm/npx
 - To setup all dev dependencies, you can run the following commands or the install.cmd file (explained further below).

    `dotnet tool restore` (to restore local dotnet tools)

    `dotnet fake build -t setup`, which will take care of installing necessary certificates and loopback exempts for your browser. Here are the steps if you want to execute them by yourself:

    - connections from excel to localhost need to be via https, so you need a certificate and trust it. [office-addin-dev-certs](https://www.npmjs.com/package/office-addin-dev-certs?activeTab=versions) does that for you.

        you can also use the fake build target for certificate creation and installation by using `dotnet fake build -t createdevcerts` in the project root (after restoring tools via `dotnet tool restore`).

        This will use office-addin-dev-certs to create the necessary certificates, and open the installation dialogue for you:

        ![File](docsrc/files/img/certificate.png)

        installing this ca certificate under your __trusted root certification authorities__ will enable you to use https via localhost.

        The paths to these certificates are then added to your webpack.config file.

     - You may need a loopback exemption for Edge/IE (whatever is run in your excel version): 

        `CheckNetIsolation LoopbackExempt -a -n="microsoft.win32webviewhost_cw5n1h2txyewy"`

### use install.cmd

The install.cmd executes several console commands for one of which it needs adminstratorial rights (dotnet fake build -t setup) to install the certificate mentioned above.
Open powershell as adminstrator and navigate to the Spawn-folder ```cd your\folder\path\Spawn``` then use ```.\install.cmd``` to initialize the setup.
While running a installation dialogue for the certificate will open and can be handled as described above.

### Project Decription

This project uses the [SAFE Stack](https://github.com/SAFE-Stack) to create a website that uses [office.js](https://github.com/OfficeDev/office-js) to interop with Word.

The file [OfficeJS.fs](src/Client/OfficeJS.fs) contains Fable bindings autogenerated from the office.js typescript definitions.

to debug the AddIn locally, use the build target `OfficeDebug`:

`fake build -t OfficeDebug`

this will launch an Word instance with the add-in sideloaded.
The app runs at localhost:3000 for client and localhost:5000 for the server.

You can use [EdgeDevToolsPreview](https://www.microsoft.com/en-us/p/microsoft-edge-devtools-preview/9mzbfrmz0mnj?activetab=pivot:overviewtab) for debugging the browser add-in (console, html) _or_ you can use Word online for development:

 - open Word online in your favorite browser
 - create a blank new workbook (or use any workbook, but be warned that you can't undo stuff done via Office.js) 
 - Go to Insert > Add-Ins > Upload my Add--Ins and upload the manifest.xml file contained in this repo.
 - You will now have the full debug experience in your browser dev tools.

Alternatively, you can debug all functionality that does not use Excel Interop in your normal browser (the app runs on port 3000 via https)

### Office.js

1. install ts2fable, needs typescript declaration files from office.js
2. donwload office.js and create fs files from it
3. adjust webpack and index.html

Necessary links:
- https://github.com/fable-compiler/ts2fable
- https://www.npmjs.com/package/@microsoft/office-js
- use office ja as external: [Swate/webpack.config.js at developer · nfdi4plants/Swate (github.com)](https://github.com/nfdi4plants/Swate/blob/developer/webpack.config.js#L117)
- [Swate/index.html at developer · nfdi4plants/Swate (github.com)](https://github.com/nfdi4plants/Swate/blob/developer/src/Client/index.html#L5)


RELEASE_NOTES.md
### 0.1.0+eab94b2 (Released 2021-3-12)
* Additions:
    * latest commit #eab94b2
    * [[#aca9810](https://github.com/nfdi4plants/Spawn/commit/aca9810d7283a178c75efd9a645d38a7d18adf5e)] Update base web structure :hammer:
    * [[#c63f8ac](https://github.com/nfdi4plants/Spawn/commit/c63f8acf18b1ee79d9bf19802ba17daf78ed83fd)] Update issue templates
    * [[#3cd5fb8](https://github.com/nfdi4plants/Spawn/commit/3cd5fb8ad32a4760c83bb0fe0391efbf3b2837c9)] Finalize minimal structure :sparkles:
    * [[#7cca87c](https://github.com/nfdi4plants/Spawn/commit/7cca87c4a85beb1304756ff0427e99e7e0a5ac27)] Start working on access to SwateDB
    * [[#56fae49](https://github.com/nfdi4plants/Spawn/commit/56fae49a7afa609d046c132d819bd87c35c0e133)] Create scalable building block logic :sparkles:
    * [[#f35bbe1](https://github.com/nfdi4plants/Spawn/commit/f35bbe16f6c110b96992c6720c988ee0b04eefea)] Add visual feedback when term is selected.
    * [[#ec8fb12](https://github.com/nfdi4plants/Spawn/commit/ec8fb12ebdc2f5e3197b9908693f794d719c5d6b)] Add parent directed search.

### 0.0.1+cfa6495 (Released 2021-3-9)
* Additions:
    * latest commit #cfa6495
    * [[#cfa6495](https://github.com/nfdi4plants/Spawn/commit/cfa6495a6dafda2ca19cada700e4c8e8bd7b072c)] Update release note generation :hammer:
    * [[#98220ae](https://github.com/nfdi4plants/Spawn/commit/98220aecf1413e27262e45bb9710959ce0e4971a)] Initial commit :tada:

### 0.0.0 (Released 2021-1-7)
* Additions:
    * Initial set up for RELEASE_Notes.md


webpack.config.js
// Template for webpack.config.js in Fable projects
// Find latest version in https://github.com/fable-compiler/webpack-config-template

// In most cases, you'll only need to edit the CONFIG object (after dependencies)
// See below if you need better fine-tuning of Webpack options

// Dependencies. Also required: core-js, fable-loader, fable-compiler, @babel/core,
// @babel/preset-env, babel-loader, sass, sass-loader, css-loader, style-loader, file-loader, resolve-url-loader
var path = require('path');
var webpack = require('webpack');
var HtmlWebpackPlugin = require('html-webpack-plugin');
var CopyWebpackPlugin = require('copy-webpack-plugin');
var MiniCssExtractPlugin = require('mini-css-extract-plugin')

var CONFIG = {
    // The tags to include the generated JS and CSS will be automatically injected in the HTML template
    // See https://github.com/jantimon/html-webpack-plugin
    indexHtmlTemplate: './src/Client/index.html',
    fsharpEntry: './src/Client/Client.fsproj',
    cssEntry: './src/Client/style.scss',
    outputDir: './deploy/public',
    assetsDir: './src/Client/public',
    devServerPort: 3000,
    // When using webpack-dev-server, you may need to redirect some calls
    // to a external API server. See https://webpack.js.org/configuration/dev-server/#devserver-proxy
    devServerProxy: {
        // redirect requests that start with /api/ to the server on port 8085
        '/api/**': {
            target: 'http://localhost:5000',
            secure: false,
            changeOrigin: true,
            ignorePath: false
           },
        // redirect websocket requests that start with /socket/ to the server on the port 8085
        '/socket/**': {
            target: 'http://localhost:5000',
            ws: true,
            secure: false,
            changeOrigin: true,
            ignorePath: false
           }
       },
    babel: {
        presets: [
            ['@babel/preset-env', {
                modules: false,
                // This adds polyfills when needed. Requires core-js dependency.
                // See https://babeljs.io/docs/en/babel-preset-env#usebuiltins
                // Note that you still need to add custom polyfills if necessary (e.g. whatwg-fetch)
                useBuiltIns: 'usage',
                corejs: 3
            }]
        ],
    }
}

// If we're running the webpack-dev-server, assume we're in development mode
var isProduction = !process.argv.find(v => v.indexOf('webpack-dev-server') !== -1);
var environment = isProduction ? 'production' : 'development';
process.env.NODE_ENV = environment;
console.log('Bundling for ' + environment + '...');

// The HtmlWebpackPlugin allows us to use a template for the index.html page
// and automatically injects <script> or <link> tags for generated bundles.
var commonPlugins = [
    new HtmlWebpackPlugin({
        filename: 'index.html',
        template: resolve(CONFIG.indexHtmlTemplate)
    })
];

module.exports = {
    // In development, split the JavaScript and CSS files in order to
    // have a faster HMR support. In production bundle styles together
    // with the code because the MiniCssExtractPlugin will extract the
    // CSS in a separate files.
    entry: isProduction ? {
        app: [resolve(CONFIG.fsharpEntry), resolve(CONFIG.cssEntry)]
    } : {
            app: resolve(CONFIG.fsharpEntry),
            style: resolve(CONFIG.cssEntry)
    },
    // Add a hash to the output file name in production
    // to prevent browser caching if code changes
    output: {
        path: resolve(CONFIG.outputDir),
        filename: isProduction ? '[name].[hash].js' : '[name].js'
    },
    mode: isProduction ? 'production' : 'development',
    devtool: isProduction ? 'source-map' : 'eval-source-map',
    optimization: {
        splitChunks: {
            chunks: 'all'
        },
    },
    // Besides the HtmlPlugin, we use the following plugins:
    // PRODUCTION
    //      - MiniCssExtractPlugin: Extracts CSS from bundle to a different file
    //          To minify CSS, see https://github.com/webpack-contrib/mini-css-extract-plugin#minimizing-for-production
    //      - CopyWebpackPlugin: Copies static assets to output directory
    // DEVELOPMENT
    //      - HotModuleReplacementPlugin: Enables hot reloading when code changes without refreshing
    plugins: isProduction ?
        commonPlugins.concat([
            new MiniCssExtractPlugin({ filename: 'style.[name].[hash].css' }),
            new CopyWebpackPlugin({ patterns: [{ from: resolve(CONFIG.assetsDir) }]}),
        ])
        : commonPlugins.concat([
            new webpack.HotModuleReplacementPlugin(),
        ]),
    externals: {
        officejs: 'Office.js',
    },
    resolve: {
        // See https://github.com/fable-compiler/Fable/issues/1490
        symlinks: false
    },
    // Configuration for webpack-dev-server
    devServer: {
        publicPath: '/',
        contentBase: resolve(CONFIG.assetsDir),
        host: '0.0.0.0',
        port: CONFIG.devServerPort,
        https: {
            key: "{USERFOLDER}/.office-addin-dev-certs/localhost.key",
            cert: "{USERFOLDER}/.office-addin-dev-certs/localhost.crt",
            ca: "{USERFOLDER}/.office-addin-dev-certs/ca.crt"
        },
        proxy: CONFIG.devServerProxy,
        hot: true,
        inline: true
    },
    // - fable-loader: transforms F# into JS
    // - babel-loader: transforms JS to old syntax (compatible with old browsers)
    // - sass-loaders: transforms SASS/SCSS into JS
    // - file-loader: Moves files referenced in the code (fonts, images) into output folder
    module: {
        rules: [
            {
                test: /\.fs(x|proj)?$/,
                use: {
                    loader: 'fable-loader',
                    options: {
                        babel: CONFIG.babel
                    }
                }
            },
            {
                test: /\.js$/,
                exclude: /node_modules/,
                use: {
                    loader: 'babel-loader',
                    options: CONFIG.babel
                },
            },
            {
                test: /\.(sass|scss|css)$/,
                use: [
                    isProduction
                        ? MiniCssExtractPlugin.loader
                        : 'style-loader',
                    'css-loader',
                    {
                        loader: 'sass-loader',
                        options: { implementation: require('sass') }
                    }
                ],
            },
            {
                test: /\.(png|jpg|jpeg|gif|svg|woff|woff2|ttf|eot)(\?.*)?$/,
                use: ['file-loader']
            }
        ]
    }
};

function resolve(filePath) {
    return path.isAbsolute(filePath) ? filePath : path.join(__dirname, filePath);
}
